docs/architecture-decisions/decouple-flag-source-and-set.md

```diff
@@ -0,0 +1,77 @@
+---
+status: proposed
+author: @tangenti
+created: 2025-06-16
+updated: 2025-06-16
+---
+
+# Decouple flag sync sources and flag sets
+
+The goal is to support dynamic flag sets for flagd providers and decouple sources and flag sets.
+
+## Background
+
+Flagd daemon syncs flag configurations from multiple sources. A single source provides a single config, which has an optional flag set ID that may or may not change in the following syncs of the same source.
+
+The in-process provider uses `selector` to specify the desired source.  In order to get a desired flag set, a provider has to stick to a source that provides that flag set. In this case, the flagd daemon cannot remove a source without breaking the dependant flagd providers.
+
+Assumptions of the current model
+
+- `flagSetId`s must be unique across different sources or the configuration is considered invalid.
+- In-process providers request at most one flag set.
+
+## Requirements
+
+- Flagd daemon can remove a source without breaking in-process providers that depend on the flag set the source provides.
+- In-process providers can select based on flag sets.
+- No breaking changes for the current usage of `selector`
+
+## Proposal
+
+### API change
+
+#### Flag Configuration Schema
+
+Add an optional field `flagsetID` under `flag` or `flag.metadata`. The flag set ID cannot be specified if a flag set ID is specified for the config.
```

Comment:
```suggestion
Add an optional field `flagsetId` under `metadata` or `flags.{flagKey}.metadata`. The flag set ID cannot be specified if a flag set ID is specified for the config.
```

Feel free to accept or not, but this is what we currently support in our config, it just doesn't have most of the behavior you describe, but I think such behavior could be added using these existing fields. See [playground](https://flagd.dev/playground/?scenario-name=Flag+metadata).

Reply:
I'm aware of that, and this is the config-level flag set ID that I mentioned in the second sentence.

It will be used as the flag set ID for every flag in the config, if the flag level flag set ID is not set.